### PR TITLE
[Mailer][FrameworkBundle] Add a way to use a custom transport

### LINKED ISSUE #31385: [Mailer] Allow create custom transports and register DSN for it
For the now [there is no extension](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Mailer/Transport.php) point for registering custom DSN for adding own Mailer Transport. Would be nice allow create custom transports like already done in Messenger.

### PR DESCRIPTION
**Pull request by kevin-verschaeve**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no   
| Deprecations? | no
| Tests pass?   | todo
| Fixed tickets | #31385 
| License       | MIT
| Doc PR        | todo

Add a "custom" provider, to allow using a custom transport class.

`MAILER_DSN=http://user:pass@custom`

can fix https://github.com/symfony/symfony/issues/31385

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
please rebase for 4.4

**Comment by kevin-verschaeve** [comment]
Branch changed to 4.4

File: src/Symfony/Component/Mailer/Transport.php [hunk 1]
```diff
@@ -169,6 +169,12 @@ private static function createTransport(string $dsn, EventDispatcherInterface $d
                 }
 
                 throw new LogicException(sprintf('The "%s" scheme is not supported for mailer "%s".', $parsedDsn['scheme'], $parsedDsn['host']));
+            case 'custom':
+                if (!$customTransport) {
+                    throw new \LogicException('You must specify the transport class when using the "custom" provider.');
+                }
+
+                return new $customTransport($user, $pass, $query, $client, $dispatcher, $logger);
```
Comment (ajgarlag) [inline src/Symfony/Component/Mailer/Transport.php]: `$customTransport` is an instance of `TransportInterface`, not a class name.
```suggestion
                return $customTransport;
```

Comment (kevin-verschaeve) [reply src/Symfony/Component/Mailer/Transport.php]: Hmm, yes you are right (I changed it just befor pushing), but what I wanted to do is to keep the data from the dsn (user, pass, and query), so I should keep the new and arguments, put do it an other way in the container extension

Comment (kevin-verschaeve) [reply src/Symfony/Component/Mailer/Transport.php]: I have edited, what do you think of this: https://github.com/symfony/symfony/pull/31931/commits/0c1a607908fcc8dc4e890491365e4a03f6a45530 ?

File: src/Symfony/Component/Mailer/Transport.php [hunk 2]
```diff
@@ -49,16 +49,16 @@ public static function fromDsn(string $dsn, EventDispatcherInterface $dispatcher
...
 
             return new Transport\RoundRobinTransport($transports);
         }
 
-        return self::createTransport($dsn, $dispatcher, $client, $logger);
+        return self::createTransport($dsn, $dispatcher, $client, $logger, $customTransportClass);
     }
 
-    private static function createTransport(string $dsn, EventDispatcherInterface $dispatcher = null, HttpClientInterface $client = null, LoggerInterface $logger = null): TransportInterface
+    private static function createTransport(string $dsn, EventDispatcherInterface $dispatcher = null, HttpClientInterface $client = null, LoggerInterface $logger = null, TransportInterface $customTransportClass = null): TransportInterface
```
Comment (ajgarlag) [inline src/Symfony/Component/Mailer/Transport.php]:
```suggestion
    private static function createTransport(string $dsn, EventDispatcherInterface $dispatcher = null, HttpClientInterface $client = null, LoggerInterface $logger = null, string $customTransportClass = null): TransportInterface
```

File: src/Symfony/Component/Mailer/Transport.php [hunk 3]
```diff
@@ -169,6 +169,12 @@ private static function createTransport(string $dsn, EventDispatcherInterface $d
                 }
 
                 throw new LogicException(sprintf('The "%s" scheme is not supported for mailer "%s".', $parsedDsn['scheme'], $parsedDsn['host']));
+            case 'custom':
+                if (!$customTransportClass) {
+                    throw new \LogicException('You must specify the transport class when using the "custom" provider.');
+                }
+
+                return new $customTransportClass($user, $pass, $query, $client, $dispatcher, $logger);
```
Comment (ajgarlag) [inline src/Symfony/Component/Mailer/Transport.php]: The `TransportInterface` does not include the class constructor, so you don't know here how to build the custom constructor class.

Also you'd have to check if `$customTrasportClass` implements `TransportInterface`

Comment (kevin-verschaeve) [reply src/Symfony/Component/Mailer/Transport.php]: Even if the interface does not include the constructor, we can pass the more arguments we have, to enable the developer to do anything he needs. Also, the developer can implements the TransportInterface or extends one of the Abstract*Transport classes.

I have added the class implements check

**Comment by fabpot** [comment]
This implementation would only allow one custom DSN, but we want third-parties to be able to provide their own DSN, so we need multiple possible custom DSNs.

**Pull request closed without merging**